src/mixins/interfaces.ts

```diff
@@ -20,3 +21,38 @@ export interface Parent {
 	children: Map<string, Child> | List<Child>;
 	append(child: Child | Child[]): Handle;
 }
+
+/**
+ * Registry to (asynchronously) get instances by their ID.
+ */
+export interface Registry<T> {
+	/**
+	 * Asynchronously get an instance by its ID.
+	 *
+	 * @param id Identifier for the instance that is to be retrieved
+	 * @return A promise for the instance. The promise rejects if no instance was found.
+	 */
+	get(id: string | symbol): Promise<T>;
+
+	/**
+	 * Look up the identifier for which the given value has been registered.
+	 *
+	 * Throws if the value hasn't been registered.
+	 *
+	 * @param value The value
+	 * @return The identifier
+	 */
+	identify(value: T): string | symbol;
+}
+
+/**
+ * Provides access to read-only registries.
+ */
+export interface RegistryProvider<T> {
+	/**
+	 * Get a registry. Different widgets expect different types of registries.
+	 *
+	 * Implementations should throw if the given type is not supported.
+	 */
+	get(type: 'actions' | 'stores' | 'widgets'): Registry<T>;
```

Comment:
@kitsonk rather than making typing assumptions for actions, stores and widgets, this approach lets each mixin define what kind of registry provider it receives. Downside though is that without using union types the mixin can get only one kind of registry out of the provider, and even if union types are used then the mixin needs to typecast each registry…

Perhaps each type could have its interface, and union typing to combine provider functionality for mixins that need multiple registries? I'm also not sure whether the current approach precludes stateful children and listeners mixins to be combined… will investigate.